### PR TITLE
POST /signup

### LINKED ISSUE #14: POST | /signup
relates #5

#### Acceptance Criteria
- [ ] Validation
- [ ] DB query
- [ ] Response
- [ ] Testing
---

#### Validation
``` js
{
 username: yup.string().alphanum().min(5).required(),
 email: yup.string().email().required(),
 password: yup.string().min(8).required(),
 confirmPassword: yup.string().oneOf([yup.ref('password'), null], 'Passwords must match')
}
```
---

#### DB query
``` js
INSERT INTO users(user_name,email,password) VALUES(user_name,email,password);
```

#### Response
- Success
```js
{ 
  StatusCode: 200,
  message :"Sign up successful"
}
```

- Failed
```js
{
statusCode:401, 
message:"You are already registered"
}
```

#### Testing
to be a successful test
- status === 200
- data.length !== 0

### PR DESCRIPTION
**Pull request by Israa91**
relates #14

### CODE REVIEW

File: server/controllers/users/addUsers.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+const { sign } = require('../../utils/jwt');
+const { addUsersQuery, checkEmail } = require('../../database/queries');
+const { boomify } = require('../../utils');
+const { addUserSchema } = require('../../utils/validation');
+
+const addUsers = async (req, res, next) => {
+  try {
+    const { userName, email, password, confirmPassword } = req.body;
+    await addUserSchema.validate(
+      { userName, email, password, confirmPassword },
```
Comment (Mu7ammadAbed) [inline server/controllers/users/addUsers.js:11]:
```suggestion
    const { userName, email, password } = await addUserSchema.validate(
      req.body,
```

File: server/controllers/users/addUsers.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+      {
+        abortEarly: false,
+      }
+    );
+    const { rows: user } = await checkEmail({ email });
+    if (user[0]) {
+      next(boomify(401, 'Login Error', 'You are registered'));
```
Comment (Mu7ammadAbed) [inline server/controllers/users/addUsers.js:19]: If the user exists this message will be sent and continue to the line after it so you need to wrap your function into `if-else` conditions or to return this line

```suggestion
      return next(boomify(401, 'Login Error', 'You are registered'));
```

File: server/controllers/users/addUsers.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    const hashedPassword = await hash(password, 10);
+    const { rows: data } = await addUsersQuery(userName, email, hashedPassword);
+    const token = sign({ userId: data[0].id });
```
Comment (Mu7ammadAbed) [inline server/controllers/users/addUsers.js:23]:
```suggestion
    const { id: userId, is_admin: isAdmin } = data[0];
    const token = sign({ userId, isAdmin });
```

File: server/database/queries/users/addUsersQuery.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+const connection = require('../../data/connection');
+
+const addUsersQuery = (userName, email, password) => {
+  const sql = {
+    text:
+      'INSERT INTO users(user_name,email,password) VALUES($1 ,$2 ,$3) returning *;',
```
Comment (Mu7ammadAbed) [inline server/database/queries/users/addUsersQuery.js:6]: The `*` will return everything even the hashed password which is something we don't want
```suggestion
      'INSERT INTO users(user_name,email,password) VALUES($1 ,$2 ,$3) returning id, user_name, email;',
```

**Review by Mu7ammadAbed — CHANGES_REQUESTED** [review]

**Review by Mu7ammadAbed — APPROVED** [review]

**Review by Osama-you — APPROVED** [review]

**Review by Alaalser — APPROVED** [review]

**Pull request merged**